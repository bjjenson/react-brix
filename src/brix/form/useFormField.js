import { actions } from './fieldReducer'
/**
 * @param  fieldArgs { import("../form").IFormFieldArgs}
 * @returns {import("../form").IFormField}
 */
export const useFormField = (state, dispatch, fieldArgs = {}) => {
  const requiredMessage = fieldArgs.requiredMessage || 'Required'

  const tryValidate = (value, touched) => {
    let result
    if (value && fieldArgs.validate) {
      result = fieldArgs.validate(value, fieldArgs.name, state.get('getAllValues'))
    }

    if (!Boolean(result) && !fieldArgs.optional && touched && value === '') {
      result = requiredMessage
    }

    if (touched) {
      dispatch(actions.validationResult(fieldArgs.name, Boolean(result), result))
    }
    return !Boolean(result)
  }

  const setValue = v => {
    dispatch(actions.updateValue(fieldArgs.name, v))
    tryValidate(v, state.getIn(['current', 'touched']))
  }

  const onChange = ({ target }) => {
    const value = fieldArgs.valueFromTarget ? fieldArgs.valueFromTarget(target) : target.value
    const coercedValue = fieldArgs.normalize ? fieldArgs.normalize(value) : value
    setValue(coercedValue)
  }

  const onBlur = () => {
    dispatch(actions.touched(fieldArgs.name))
    tryValidate(state.getIn(['current', 'value']), true)
  }

  const setValidationResult = result => {
    dispatch(actions.validationResult(fieldArgs.name, true, result))
  }

  const validate = () => {
    return tryValidate(state.getIn(['current', 'value']), true)
  }

  return {
    props: {
      error: state.getIn(['current', 'error']),
      helperText: state.getIn(['current', 'helperText']),
      label: state.getIn(['initial', 'label']),
      value: state.getIn(['current', 'value']),
      // handlers
      onBlur,
      onChange,
    },
    setValidationResult,
    setValue,
    validate,
  }
}
