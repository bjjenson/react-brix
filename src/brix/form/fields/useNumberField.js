import { useFormField } from '../useFormField'

export const useNumberField = (args = {}, initial) => {
  let initialValue
  if (initial !== undefined) {
    initialValue = initial
  } else if (args.value !== undefined) {
    initialValue = args.value
  } else {
    initialValue = 0
  }

  return useFormField(initialValue, args)
}
