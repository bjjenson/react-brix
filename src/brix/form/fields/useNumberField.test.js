import { useFormField } from '../useFormField'
import { useNumberField } from './useNumberField'

jest.mock('../useFormField')

let initialArgs
beforeEach(() => {
  initialArgs = {
    value: 23,
  }

  useFormField.mockReturnValue({ prop1: 'prop1', prop2: 'prop2' })
})


test('sets initial values', () => {
  useNumberField(initialArgs)
  expect(useFormField.mock.calls[0]).toMatchSnapshot()
})

test('sets initialValue from args', () => {
  useNumberField(initialArgs, false)
  expect(useFormField.mock.calls[0]).toMatchSnapshot()
})

test('defaults value to 0', () => {
  useNumberField()
  expect(useFormField.mock.calls[0]).toMatchSnapshot()
})

test('returns all props needed', () => {
  const actual = useNumberField(initialArgs)
  expect(actual).toMatchSnapshot()
})
