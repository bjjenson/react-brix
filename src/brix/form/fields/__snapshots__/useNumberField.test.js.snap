// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults value to 0 1`] = `
Array [
  0,
  Object {},
]
`;

exports[`returns all props needed 1`] = `
Object {
  "prop1": "prop1",
  "prop2": "prop2",
}
`;

exports[`sets initial values 1`] = `
Array [
  23,
  Object {
    "value": 23,
  },
]
`;

exports[`sets initialValue from args 1`] = `
Array [
  false,
  Object {
    "value": 23,
  },
]
`;
